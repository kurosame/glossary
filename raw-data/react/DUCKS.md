## category

react

## titles

Ducks
Redux

## description

Redux のデザインパターン  
Action や Reducer を modules にパッケージする

vuejs-boilerplate でも使ってる気がする  
1 ファイルにはしてないけど、actions, getters, mutations, state を import して modules でパッケージして export してる  
そして Store 作るときに modules をセットしている
