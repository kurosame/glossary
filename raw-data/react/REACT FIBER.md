## category

react

## titles

React Fiber
react-fiber

## description

Begin, Complete, Commit のフェーズがある

### Begin, Complete

Component 単位で実行される  
componentWillMount, componentWillReceiveProps, shouldComponentUpdate, componentWillUpdate  
非同期レンダリングでは中断・再開により、複数回ライフサイクル関数が呼ばれる可能性がある

### Commit

全ての Fiber が完了した時に実行される  
DOM 更新などの副作用が発生する処理はここでまとめて反映される  
componentDidMount, componentDidUpdate, componentWillUnmount

### Time Slicing

Fiber により画面のレンダリングが非同期になると、Fiber 単位で処理の優先度を付けることができる  
例えばユーザ操作の優先度を高くすることで、画面がレンダリング中でもユーザ操作を優先して処理することが可能となる

### Suspense

非同期通信をしている間、画面のレンダリングを中断できる  
非同期通信が完了した後に、画面をレンダリングすることが可能  
render に Promise を投げて、その Promise が完了した後に render するみたいな？
