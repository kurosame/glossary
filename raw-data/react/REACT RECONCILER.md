## category

react

## titles

React Reconciler
react-reconciler

## description

VDOM と呼ばれている部分

1. ReactElement のツリーを構築
1. Component をマウント、アンマウント、更新
1. ライフサイクル関数を呼出す
1. 現在の DOM との差分を計算する
1. renderer に差分を伝える
