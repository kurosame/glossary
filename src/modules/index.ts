import { ICounterActions, ICounterState } from '@/modules/counter'
import { IWordActions, IWordState } from '@/modules/word'

export interface IActions {
  counter: ICounterActions
  word: IWordActions
}

export interface IStates {
  counter: ICounterState
  words: IWordState[]
}
