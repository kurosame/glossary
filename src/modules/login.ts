import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const SET_IS_LOGIN = 'SET_IS_LOGIN'

export interface LoginActions {
  setIsLogin: ActionFunctionAny<Action<{ isLogin: boolean }>>
}

export interface LoginState {
  isLogin: boolean
}

const initialState: LoginState = { isLogin: false }

export const { setIsLogin } = createActions<{ isLogin: boolean }>(SET_IS_LOGIN)

export const login: Reducer<
  LoginState,
  Action<{ isLogin: boolean }>
> = handleActions(
  {
    [SET_IS_LOGIN]: (
      _state: LoginState,
      action: Action<{ isLogin: boolean }>
    ): LoginState => {
      return { isLogin: action.payload.isLogin }
    }
  },
  initialState
)
