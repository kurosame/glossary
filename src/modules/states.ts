import type { LoginState } from '@/modules/login'
import type { WordState } from '@/modules/word'

export interface States {
  login: LoginState
  words: WordState[]
}
