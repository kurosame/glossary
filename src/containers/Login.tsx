import type firebase from 'firebase'
import React, { SetStateAction, useEffect, useState } from 'react'
import StyledFirebaseAuth from 'react-firebaseui/StyledFirebaseAuth'
import { useDispatch, useSelector } from 'react-redux'
import type { Dispatch } from 'redux'
import firebaseApp from '@/firebase/index'
import uiConfig from '@/firebase/ui-config'
import { SET_IS_LOGIN } from '@/modules/login'
import type { States } from '@/modules/states'

const useSetUser = (su: React.Dispatch<SetStateAction<firebase.User | null>>): void =>
  useEffect(() => firebaseApp.auth().onAuthStateChanged(u => su(u)))

const useSetIsLogin = (u: firebase.User | null): void => {
  const dispatch = useDispatch<Dispatch>()
  useEffect(() => {
    dispatch({ type: SET_IS_LOGIN, payload: { isLogin: !!u } })
  }, [u, dispatch])
}

const Login: React.FC = () => {
  const [user, setUser] = useState<firebase.User | null>(null)
  const isLogin = useSelector<States, boolean>(s => s.login.isLogin)

  useSetUser(setUser)
  useSetIsLogin(user)

  if (isLogin) return null
  return <StyledFirebaseAuth uiConfig={uiConfig} firebaseAuth={firebaseApp.auth()} />
}

export default Login
