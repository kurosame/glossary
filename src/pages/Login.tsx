import React from 'react'
import StyledFirebaseAuth from 'react-firebaseui/StyledFirebaseAuth'
import { connect } from 'react-redux'
import { bindActionCreators, Dispatch } from 'redux'
import uiConfig from '@/firebase/ui-config'
import firebase from '@/firebase/index'
import { States } from '@/modules/states'
import { LoginActions, LoginState, setIsLogin } from '@/modules/login'

interface Props {
  state: { login: LoginState }
  actions: LoginActions
  store?: unknown
}

class Login extends React.PureComponent<Props> {
  componentDidMount(): void {
    firebase
      .auth()
      .onAuthStateChanged(user =>
        this.props.actions.setIsLogin({ isLogin: !!user })
      )
  }

  public render(): JSX.Element | null {
    if (this.props.state.login.isLogin) return null
    return (
      <StyledFirebaseAuth uiConfig={uiConfig} firebaseAuth={firebase.auth()} />
    )
  }
}

export default connect(
  (states: States) => ({ state: { login: states.login } }),
  (dispatch: Dispatch) => ({
    actions: {
      setIsLogin: bindActionCreators(setIsLogin, dispatch)
    }
  })
)(Login)
