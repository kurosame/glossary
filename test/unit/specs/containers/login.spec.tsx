import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import Login from '@/containers/Login'
import firebaseApp from '@/firebase/index'
import { SET_IS_LOGIN } from '@/modules/login'
import { States } from '@/modules/states'
import { render, cleanup, RenderResult } from '@testing-library/react'

let mockDispatch: jest.Mock
let spyFirebaseAuth: jest.SpyInstance
let wrapper: (isLogin: boolean) => RenderResult
beforeEach(() => {
  mockDispatch = jest.fn()
  spyFirebaseAuth = jest.spyOn(firebaseApp.auth(), 'onAuthStateChanged')
  wrapper = (isLogin: boolean): RenderResult => {
    const store = configureStore<Pick<States, 'login'>>()({
      login: { isLogin }
    })
    store.dispatch = mockDispatch
    return render(
      <Provider store={store}>
        <Login />
      </Provider>
    )
  }
})
afterEach(() => {
  cleanup()
  jest.restoreAllMocks()
})

test('Call `useSetUser`', () => {
  wrapper(true)

  expect(spyFirebaseAuth).toBeCalled()
})

test('Call `useSetIsLogin`', () => {
  wrapper(true)

  expect(mockDispatch).toBeCalled()
  expect(mockDispatch.mock.calls[0][0]).toEqual({
    type: SET_IS_LOGIN,
    payload: { isLogin: false }
  })
  expect(mockDispatch.mock.calls[1]).toBeUndefined()
})

test('Render DOM if `isLogin` is false', async () => {
  // FirebaseUI for React is not supported by Jest(jsdom)
  // https://github.com/firebase/firebaseui-web/issues/636
  // expect(wrapper(false).container.querySelector('div')).not.toBeNull()
})

test('Not render DOM if `isLogin` is true', () => {
  expect(wrapper(true).container.querySelector('div')).toBeNull()
})

test('Match the snapshot', () => {
  // FirebaseUI for React is not supported by Jest(jsdom)
  // https://github.com/firebase/firebaseui-web/issues/636
  // expect(wrapper(false).asFragment()).toMatchSnapshot()
})
