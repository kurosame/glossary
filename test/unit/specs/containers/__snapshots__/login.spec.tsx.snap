// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot 1`] = `
<DocumentFragment>
  <div
    id="firebaseui_container"
  />
</DocumentFragment>
`;

exports[`Match the snapshot The isLogin is false 1`] = `"<div id=\\"firebaseui_container\\"></div>"`;

exports[`Match the snapshot The isLogin is true 1`] = `""`;
